Core/TransactionHandler.php

```diff
@@ -0,0 +1,314 @@
+            }
+
+            return $this->_getErrorMessage(Helper::translate('text_generic_error') . '<br>' . $sErrors);
+        }
+
+        if ($oResponse instanceof SuccessResponse) {
+            // create new transaction entry in the database
+            $this->_handleNotificationSuccess($oBackendService, $oResponse);
+
+            // also update the order state in the database
+            $sUpdatedOrderState = $oBackendService->getOrderState($oResponse->getTransactionType());
+            $oOrder->oxorder__wdoxidee_orderstate = new Field($sUpdatedOrderState);
+            $oOrder->save();
+        }
+
+        return $this->_getSuccessMessage();
+    }
+
+    /**
+     * Handles the success response from the post-processing operation.
+     * A new transaction entry is added to the database.
+     *
+     * @param BackendService  $oBackendService
+     * @param SuccessResponse $oResponse
+     *
+     * @return null
+     *
+     * @since 1.0.1
+     */
+    private function _handleNotificationSuccess($oBackendService, $oResponse)
+    {
+        $aData = $oResponse->getData();
+        $oUtilsDate = Registry::getUtilsDate();
+        $sConvertedTimestamp = $oUtilsDate->formatDBTimestamp($oUtilsDate->formTime($aData['completion-time-stamp']));
+
+        // find root transaction ID
+        $sRootTransactionId = $this->_getRootTransactionId($oResponse->getParentTransactionId());
+
+        $oOrder = oxNew(Order::class);
+        if (!$oOrder->loadWithTransactionId($sRootTransactionId)) {
+            $this->_oLogger->error('No order found for transactionId: ' . $oResponse->getParentTransactionId());
+            return;
+        }
+
+        // transaction amount
+        $fAmount = $oResponse->getRequestedAmount()->getValue();
+
+        // determine if transaction state is closed
+        $sTransactionState = $oBackendService->isFinal($oResponse->getTransactionType()) ?
+                Transaction::STATE_CLOSED : $aData['transaction-state'];
+
+        // create an array with the properties to be saved in the transaction database entry
+        $aTransactionProps = array(
+            'ordernumber' => $oOrder->oxorder__oxordernr->value,
+            'orderid' => $oOrder->oxorder__oxid->value,
+            'transactionid' => $oResponse->getTransactionId(),
+            'parenttransactionid' => $oResponse->getParentTransactionId(),
+            'requestid' => $oResponse->getRequestId(),
+            'action' => $oOrder->getOrderPayment()->oxpayments__wdoxidee_transactionaction->value,
+            'type' => $oResponse->getTransactionType(),
+            'state' => $sTransactionState,
+            'amount' => $fAmount,
+            'currency' => $oResponse->getRequestedAmount()->getCurrency(),
+            'responsexml' => base64_encode($oResponse->getRawData()),
+            'date' => $sConvertedTimestamp
+        );
+
+        // add a transaction database entry from properties array
```

Comment:
We can omit this comment as it is obvious enough.

Reply:
👍